src/System.Net.Requests/src/System/Net/FtpWebRequest.cs

```diff
@@ -0,0 +1,2544 @@
+                                {
+                                    m_RemainingTimeout = Timeout - (int)((DateTime.UtcNow - m_StartTime).TotalMilliseconds);
+                                    if(m_RemainingTimeout <= 0){
+                                        throw;
+                                    }
+                                }
+                            }
+                            continue;
+                        }
+                        throw;
+                    }
+                    // no retry needed
+                    break;
+                }
+            } catch (WebException webException) {
+                //if this was a timeout, throw a timeout exception
+                IOException ioEx = webException.InnerException as IOException;
+                if(ioEx != null){
+                    SocketException sEx = ioEx.InnerException as SocketException;
+                     if(sEx != null){
+                        if (sEx.SocketErrorCode == SocketError.TimedOut) {
+                            SetException(new WebException(SR.net_timeout, WebExceptionStatus.Timeout));
+                        }
+                    }
+                }
+
+                SetException(webException);
+            }
+            catch (Exception exception) {
+                SetException(exception);
+            }
+        }
+
+        private Exception TranslateConnectException(Exception e)
+        {
+            SocketException se = e as SocketException;
+            if (se != null)
+            {
+                if (se.SocketErrorCode == SocketError.HostNotFound)
+                {
+                    return new WebException(SR.net_webstatus_NameResolutionFailure, WebExceptionStatus.NameResolutionFailure);
+                }
+                else
+                {
+                    return new WebException(SR.net_webstatus_ConnectFailure, WebExceptionStatus.ConnectFailure);
+                }
+            }
+
+            // Wasn't a socket error, so leave as is
+            return e;
+        }
+
+        private FtpControlStream CreateConnection()
+        {
+            string hostname = m_Uri.Host;
+            int port = m_Uri.Port;
+
+            TcpClient client = new TcpClient();
+
+            try
+            {
+                if (m_Async)
+                {
+                    Task t = client.ConnectAsync(hostname, port);
+                    if (!t.IsCompleted)
+                    {
+                        t.ContinueWith((t2) =>
+                        {
+                            if (t2.IsFaulted)
+                            {
+                                AsyncRequestCallback(TranslateConnectException(t2.Exception));
+                            }
+                            else
+                            {
+                                AsyncRequestCallback(new FtpControlStream(client));
+                            }
+                        });
```

Comment:
At a minimum it'd be good to pass TaskScheduler.Default as the taskScheduler argument to ContinueWith.  ContinueWith otherwise picks up the current scheduler, which for continuations is often desired in app-level code, but in library-level code you typically want to avoid unnecessarily scheduling work back to the caller's context/scheduler, in particular as it can inadvertently lead to deadlocks if they end up blocking waiting for the async operation to complete.

Separately, I'm assuming perf isn't much of an issue here, but if it is, there are some improvements that can be made.  For example, the delegate passed to ContinueWith is closing over state used even on the non-async path, which will result in additional allocations even if this isn't used.  ContinueWith also has an "object state" parameter that can be used to pass state into the delegate without using a closure.